From a8ac9541eab329781e60eda646a4337e11405fe5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Giedrius=20Medzevi=C4=8Dius?= <giedrius@8devices.com>
Date: Fri, 20 Sep 2024 01:36:31 +0300
Subject: [PATCH] ath11k: monitor: fix trailing 8 bytes in captured frames

Fix an issue where monitor mode traffic from the QCN9074 radio contains
an extra 8 bytes at the end of each frame.

The bug is caused by incorrectly using `sizeof(struct hal_rx_desc)`,
a generic union structure, for the MSDU size calculation.
The correct RX descriptor size, matching the hardware configuration,
should be used, as is done in regular RX frame processing.
Upstream-Status: Pending

---
 drivers/net/wireless/ath/ath11k/dp_rx.c | 26 +++++++++++++++----------
 1 file changed, 16 insertions(+), 10 deletions(-)

diff --git a/drivers/net/wireless/ath/ath11k/dp_rx.c b/drivers/net/wireless/ath/ath11k/dp_rx.c
index fb426195a3f0..d6f462b0cefd 100644
--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -4460,10 +4460,10 @@ int ath11k_dp_rx_pdev_alloc(struct ath11k_base *ab, int mac_id)
 	return 0;
 }
 
-static void ath11k_dp_mon_set_frag_len(u32 *total_len, u32 *frag_len)
+static void ath11k_dp_mon_set_frag_len(u32 *total_len, u32 *frag_len, u32 max_frag_len)
 {
-	if (*total_len >= (DP_RX_BUFFER_SIZE - sizeof(struct hal_rx_desc))) {
-		*frag_len = DP_RX_BUFFER_SIZE - sizeof(struct hal_rx_desc);
+	if (*total_len >= max_frag_len) {
+		*frag_len = max_frag_len;
 		*total_len -= *frag_len;
 	} else {
 		*frag_len = *total_len;
@@ -4610,7 +4610,8 @@ static u32 ath11k_dp_rx_mon_comp_ppduid(u32 msdu_ppdu_id, u32 *ppdu_id,
 
 static void ath11k_dp_mon_get_buf_len(struct hal_rx_msdu_desc_info *info,
 				      bool *is_frag, u32 *total_len,
-				      u32 *frag_len, u32 *msdu_cnt)
+				      u32 *frag_len, u32 max_frag_len,
+				      u32 *msdu_cnt)
 {
 	if (info->msdu_flags & RX_MSDU_DESC_INFO0_MSDU_CONTINUATION) {
 		if (!*is_frag) {
@@ -4618,11 +4619,11 @@ static void ath11k_dp_mon_get_buf_len(struct hal_rx_msdu_desc_info *info,
 			*is_frag = true;
 		}
 		ath11k_dp_mon_set_frag_len(total_len,
-					   frag_len);
+					   frag_len, max_frag_len);
 	} else {
 		if (*is_frag) {
 			ath11k_dp_mon_set_frag_len(total_len,
-						   frag_len);
+						   frag_len, max_frag_len);
 		} else {
 			*frag_len = info->msdu_len;
 		}
@@ -4651,6 +4652,7 @@ ath11k_dp_rx_mon_mpdu_pop(struct ath11k *ar, int mac_id,
 	u32 rx_bufs_used = 0, i = 0;
 	u32 msdu_ppdu_id = 0, msdu_cnt = 0;
 	u32 total_len = 0, frag_len = 0;
+	const u32 max_frag_len = DP_RX_BUFFER_SIZE - ar->ab->hw_params.hal_desc_sz;
 	bool is_frag, is_first_msdu;
 	bool drop_mpdu = false;
 	struct ath11k_skb_rxcb *rxcb;
@@ -4744,7 +4746,7 @@ ath11k_dp_rx_mon_mpdu_pop(struct ath11k *ar, int mac_id,
 
 			rx_desc = (struct hal_rx_desc *)msdu->data;
 
-			rx_pkt_offset = sizeof(struct hal_rx_desc);
+			rx_pkt_offset = ar->ab->hw_params.hal_desc_sz;
 			l2_hdr_offset = ath11k_dp_rx_h_msdu_end_l3pad(ar->ab, rx_desc);
 
 			if (is_first_msdu) {
@@ -4775,7 +4777,8 @@ ath11k_dp_rx_mon_mpdu_pop(struct ath11k *ar, int mac_id,
 			}
 			ath11k_dp_mon_get_buf_len(&msdu_list.msdu_info[i],
 						  &is_frag, &total_len,
-						  &frag_len, &msdu_cnt);
+						  &frag_len, max_frag_len,
+						  &msdu_cnt);
 			rx_buf_size = rx_pkt_offset + l2_hdr_offset + frag_len;
 
 			ath11k_dp_pkt_set_pktlen(msdu, rx_buf_size);
@@ -5302,6 +5305,7 @@ ath11k_dp_rx_full_mon_mpdu_pop(struct ath11k *ar,
 	u32 rx_buf_size, rx_pkt_offset, l2_hdr_offset;
 	u32 rx_bufs_used = 0, msdu_cnt = 0;
 	u32 total_len = 0, frag_len = 0, sw_cookie;
+	const u32 max_frag_len = DP_RX_BUFFER_SIZE - ar->ab->hw_params.hal_desc_sz;
 	u16 num_msdus = 0;
 	u8 rxdma_err, rbm;
 	bool is_frag, is_first_msdu;
@@ -5381,7 +5385,8 @@ ath11k_dp_rx_full_mon_mpdu_pop(struct ath11k *ar,
 
 			rx_desc = (struct hal_rx_desc *)msdu->data;
 
-			rx_pkt_offset = sizeof(struct hal_rx_desc);
+			rx_pkt_offset = ar->ab->hw_params.hal_desc_sz;
+
 			l2_hdr_offset = ath11k_dp_rx_h_msdu_end_l3pad(ar->ab, rx_desc);
 
 			if (is_first_msdu) {
@@ -5396,7 +5401,8 @@ ath11k_dp_rx_full_mon_mpdu_pop(struct ath11k *ar,
 
 			ath11k_dp_mon_get_buf_len(&msdu_list.msdu_info[i],
 						  &is_frag, &total_len,
-						  &frag_len, &msdu_cnt);
+						  &frag_len, max_frag_len,
+						  &msdu_cnt);
 
 			rx_buf_size = rx_pkt_offset + l2_hdr_offset + frag_len;
 
